server/provider/networkProvider.go

```diff
@@ -222,7 +239,27 @@ func (provider *networkProvider) doGetBlockByNonce(nonce uint64) (*data.Block, e
 		return nil, newErrCannotGetBlockByNonce(nonce, errors.New(response.Error))
 	}
 
-	return &response.Data.Block, nil
+	block = &response.Data.Block
+
+	provider.cacheBlockByNonce(nonce, block)
```

Comment:
At this point, the block may already be in cache as an effect of concurrent requests - but this should not be an issue.